

%
% NOTE -- ONLY EDIT THE .Rnw FILE!!!  The .tex file is
% likely to be overwritten.
%
\documentclass[landscape]{article}

\usepackage{amsmath,pstricks}
\usepackage[authoryear,round]{natbib}
\usepackage{hyperref}
\usepackage{sectsty}
\usepackage{landscape}
\usepackage{graphics}

\textwidth=10.9in
\textheight=6.5in
%\parskip=.3cm
\oddsidemargin=.0in
\evensidemargin=.0in
\headheight=-.3in


\newcommand{\scscst}{\scriptscriptstyle}
\newcommand{\scst}{\scriptstyle}
\newcommand{\dimfo}{\fontsize{14}{16}\selectfont}
\newcommand{\expfo}{\fontsize{18}{22}\selectfont}


\newcommand{\Rfunction}[1]{{\texttt{#1}}}
\newcommand{\Robject}[1]{{\texttt{#1}}}
\newcommand{\Rpackage}[1]{{\textit{#1}}}
\newcommand{\Rmethod}[1]{{\texttt{#1}}}
\newcommand{\Rfunarg}[1]{{\texttt{#1}}}
\newcommand{\Rclass}[1]{{\textit{#1}}}

\newcommand\bi{\begin{itemize}}
\newcommand\ei{\end{itemize}}

\textwidth=8.2in

\bibliographystyle{plainnat} 

\renewcommand{\familydefault}{\sfdefault}

\usepackage[T1]{fontenc}

\newcommand\cp{{\clearpage}}
\newcommand\sts[1]{\Huge \textbf{#1}}
\newcommand\stsh[1]{\huge \textbf{#1}}
\newcommand\stsL[1]{\Large \textbf{#1}}
\newcommand\stsl[1]{\large \textbf{#1}}

\begin{document}

\fontsize{18}{22}
\selectfont
\allsectionsfont{\sffamily}
\sectionfont{\fontfamily{phv}\fontsize{18}{22}\selectfont}
\subsectionfont{\fontfamily{phv}\fontsize{18}{22}\selectfont}
\subsubsectionfont{\fontfamily{phv}\fontsize{18}{22}\selectfont}


%\setkeys{Gin}{width=0.85\textwidth}

{
\Huge
\begin{center}
\textbf{Exploring 1000 Genomes with Bioconductor} \\
\vspace*{1cm}
Vince Carey \\ Channing Lab \\ Harvard Medical School
\end{center}
\vspace*{1.5cm}
\begin{itemize}
\item Prologue: What is an eQTL?
\item Sketches: 1000 genomes; Bioconductor
\item Imputation to the 1000 genomes SNP panel
\item Expression arrays, RNA-seq, and eQTL identification
\ei
}

\clearpage

{\sts{Prologue: What is an eQTL (expression quantitative trait locus)?}
\bi
\item Arises from a basic form of integrative genome-scale data analysis
\item On a cohort of $N$ individuals
\bi
\item SNP-chip yields allele counts for $S$ SNP, $S \approx 10^6$
\item Expression array yields mRNA abundance measures for $G$ genes, $G \approx 20000$
\ei
\item perform $G \times S$ association tests of $H_{ogs}:$ mean expression of $g$ is independent of allele count for $s$
\item the best hits are eQTL
\ei
}

\clearpage

\textbf{A chromosome-wide scan for a single gene}

\begin{center}
\setkeys{Gin}{width=0.75\textwidth}
\includegraphics{showEQTL}
\end{center}

\clearpage

\textbf{The `best SNP' discriminates mean expression}

\begin{center}
\setkeys{Gin}{width=0.75\textwidth}
\includegraphics{goodEQTL}
\end{center}

\clearpage

\textbf{Dixon 2007 Nat Genet `global map'}

\begin{center}
\setkeys{Gin}{width=0.85\textwidth}
\includegraphics{dixonSurvey}
\end{center}

\clearpage

\textbf{Why do this?  1: Mechanisms of transcriptional control}

\begin{center}
\setkeys{Gin}{width=0.85\textwidth}
\includegraphics{williamsSchem}
\end{center}

(RBH Williams et al 2007 Genome Resch)

\clearpage

{\stsL{Why do this?  2: Filtering SNP for efficient GWAS}

\bi
\item SNPs binned left to right in decreasing order of expression regulatory capacity
\item $y$ axis: proportion SNP in bin associated with macro phenotype in WTCCC
\ei

\setkeys{Gin}{width=0.95\textwidth}
\includegraphics{nicolaeEnrich}

(D Nicolae et al 2010 PLoS Genetics)
}

\clearpage

{\sts{Upshots}
\bi
\item eQTL catalogs seem useful; can efficiencies for
individual studies be gained by imputing
denser SNP panels using results of institutional deep sequencing?
\item How can higher-resolution measures of mRNA abundance
add to value from eQTL concepts: eQTL searches based on
RNA-seq/DNA-seq?
\item Under the hood, things may not be so nice...
\ei
}

\clearpage

\setkeys{Gin}{width=0.76\textwidth}
\includegraphics{withCircs}

\clearpage

{\sts{Sketch: 1000 genomes}

\begin{center}
\setkeys{Gin}{width=0.96\textwidth}
\includegraphics{flicekMap}
\end{center}

}

\clearpage

{\sts{Sketch: 1000 genomes}

\bi
\item DNA sequencing to various depths; high-level interfaces
via browsers
\item public release to aligned read level: many many BAM files
\item more tractable: SNP and variant `calls': VCF files
\item while focus is on DNA variation, availability of cell
lines permits measurement of various microscopic phenotypes
\item summary: 
\bi
\item archive of genetic sequence
\item institutional data reductions
\item resource for inference on genetic hypotheses and for methods development
\ei
\ei
}

\clearpage

{\sts{Sketch: Bioconductor}

\bi
\item open-source repository for R-based software targeting genome-scale
data analysis
\item progress to date
\bi
\item preprocessing/annotation/analysis
\item important methods support for affy and illumina expression and
genotyping arrays
\item interfaces to GEO/ArrayExpress/SRA for rapid import
\item support for high-performance GWAS and eQTL searches
\item exploit innovations in R: multicore, ``disk as RAM'', ``orchestrator''
\item efforts in sequencing: QC, annotation, analysis (particularly RNA-seq)
\item for 1000 genomes, we have ind1KG, ceu1KG
\ei
\ei
}

\clearpage

\begin{verbatim}
ceu1kg-package             package:ceu1kg              R Documentation

CEU (N=60) genotypes from 1000 genomes pilot phase I

Description:

     CEU genotypes from 1000 genomes pilot phase I (approx 8 million
     SNP); includes wellcome trust GENEVAR expression for 41
     individuals

Details:

       Package:   ceu1kg                                                  
       Version:   0.0.0                                                   
       Depends:   R (>= 2.11.1), snpMatrix (>= 1.13.1), GGBase (>= 3.9.0) 
       License:   Artistic-2.0                                            
       LazyLoad:  yes                                                     
       Built:     R 2.12.0; ; 2010-07-01 01:14:27 UTC; unix               
      
     Index:
     
     
     ceu1kg-package          60 hapmap CEU samples, 47K expression, 8mm 1000
                             genomes SNP 

     There are three basic data resources provided here.

     First, the 1000 genomes SNP calls for 60 CEU individuals were
     extracted from the pilot 1 VCF files distributed at <URL:
     ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/pilot_data/release/2010_03/pilot1/>.

     Second, metadata 'GRanges-class' instances are provided in
     chromosome-specific containers.

     Third, an 'smlSet' is provided for 41 individuals in the 1000
     genomes CEU SNP call set for whom expression data are available
     via the Sanger GENEVAR distribution (<URL:
     ftp://ftp.sanger.ac.uk/pub/genevar/CEU_parents_norm_march2007.zip>).

\end{verbatim}

\clearpage
<<lll,echo=FALSE,results=hide,eval=FALSE,results=hide,echo=FALSE>>=
library(ceu1kg)
if (!exists("ceu1KG.sml")) data(ceu1KG.sml)
if (!exists("ceu1kg")) data(ceu1kg)
<<lkd,eval=FALSE,echo=FALSE>>=
library(ceu1kg)
data(ceu1KG.sml)
<<lkddd,eval=FALSE,results=hide,echo=FALSE>>=
sapply(ceu1KG.sml,dim)
ceu1KG.sml[[1]][1:2,1:5]
as(ceu1KG.sml[[1]][1:2,1:5], "matrix")
as(ceu1KG.sml[[1]][1:2,1:5], "character")
@

%\clearpage
<<lkm,eval=FALSE,results=hide,echo=FALSE>>=
ceu1kg
dim(exprs(ceu1kg))
summary(smList(ceu1kg)[[20]])
@
\input{toughcode}
%\bi
%\item derived from the VCF representation of all calls for CEU
%\item distributed as a 700MB gzipped tabix-indexed file
%\item \texttt{GGtools:::vcf2sm} imports record-at-a-time
%decompressing on the fly
%\ei

\clearpage

{\sts{Exercise}
\bi
\item The FKBP1A eQTL finding highlighted before was derived
from HapMap phase II calls
\item Reproduce the finding based on GENEVAR expression data
and 1000 genomes calls
\ei
{\huge
\begin{verbatim}
> plot_EvG(probeId("GI_17149835-I"), rsid("rs6041750"), 
       ceu1kg[chrnum("chr20"),])
Error in .local(x, rs, ...) : rs number not found in columns of smlSet
\end{verbatim}
\bi
\item Account for revisions in dbSNP
\ei
<<doo,eval=FALSE>>=
plot_EvG(probeId("GI_17149835-I"), rsid("rs8392"), ceu1kg[chrnum("chr20"),])
@
} %huge
}

\clearpage

\includegraphics{redoEVG}

\clearpage

{\sts{Summary}
\bi
\item We have 7.7 million SNP calls on each of 60 individuals
\item For 41 individuals with widely disseminated expression data,
this has been linked to genotype
\item Modulo a change to allele coding we reproduced a phase II
HapMap eQTL finding using the 1000 genomes calls
\item We have tools to import the 1000 genomes call sets for any
population distributed through the project, or any call set in VCF format
\ei
}

\clearpage

\includegraphics{1KGVCFarch}

\clearpage

{\Huge NB: still need data/analysis models for variants beyond SNP}

\includegraphics{structuralVar}





\end{document}
